homeassistant/components/zwave_js/__init__.py

```diff
@@ -182,50 +186,77 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        await client.disconnect()
 
-    listen_task = asyncio.create_task(
-        client_listen(hass, entry, client, driver_events.ready)
-    )
-    entry_hass_data[DATA_CLIENT_LISTEN_TASK] = listen_task
     entry.async_on_unload(
         hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, handle_ha_shutdown)
     )
 
+    driver_ready_task = hass.async_create_task(driver_ready.wait())
     try:
-        await driver_events.ready.wait()
-    except asyncio.CancelledError:
-        LOGGER.debug("Cancelling start client")
-        return
+        async with asyncio.timeout(LISTEN_READY_TIMEOUT):
+            done, _ = await asyncio.wait(
+                (driver_ready_task, listen_task), return_when=asyncio.FIRST_COMPLETED
+            )
+    except asyncio.TimeoutError as err:
+        driver_ready_task.cancel()
+        listen_task.cancel()
+        raise ConfigEntryNotReady("Driver ready timed out") from err
+
+    if listen_task in done:
+        # If the listen task is already done, we need to raise ConfigEntryNotReady
+        listen_error = listen_task.exception()
+        driver_ready_task.cancel()
+        raise ConfigEntryNotReady(listen_error) from listen_error
 
-    LOGGER.info("Connection to Zwave JS Server initialized")
+    LOGGER.debug("Connection to Zwave JS Server initialized")
 
-    assert client.driver
-    async_dispatcher_send(
-        hass, f"{DOMAIN}_{client.driver.controller.home_id}_connected_to_server"
+    entry_hass_data: dict = hass.data[DOMAIN].setdefault(entry.entry_id, {})
+    entry_hass_data[DATA_CLIENT] = client
+    entry_hass_data[DATA_CLIENT_LISTEN_TASK] = listen_task
+    entry_hass_data[DATA_DRIVER_EVENTS] = driver_events = DriverEvents(hass, entry)
+
+    driver = client.driver
+    # When the driver is ready we know it's set on the client.
+    assert driver is not None
+
+    with contextlib.suppress(NotConnected):
+        # If the client isn't connected the listen task will have an exception
+        # and we'll handle the clean up below.
+        await driver_events.setup(driver)
+
+    # If the listen task is already failed, we need to raise ConfigEntryNotReady
+    if listen_task.done() and (listen_error := listen_task.exception()) is not None:
+        tasks: list[Coroutine] = [
+            hass.config_entries.async_forward_entry_unload(entry, platform)
+            for platform, task in driver_events.platform_setup_tasks.items()
+            if not task.cancel()
+        ]
+        if tasks:
+            await asyncio.gather(*tasks)
+        hass.data[DOMAIN].pop(entry.entry_id)
+        try:
+            await client.disconnect()
+        finally:
+            raise ConfigEntryNotReady(listen_error) from listen_error
+
+    # Re-attach trigger listeners.
+    # Schedule this call to make sure the config entry is loaded first.
+    hass.async_add_job(
+        async_dispatcher_send,
+        hass,
+        f"{DOMAIN}_{driver.controller.home_id}_connected_to_server",
     )
```

Comment:
This seems like it might be brittle since if the function calling this awaits before it changes the config entry state is might run too soon

Reply:
Integrations should be able to and can currently trust that their config entry is loaded when returning true here. Otherwise things will become a lot more complicated in general.

Reply:
Alternatively we can allow integrations to listen for the `SIGNAL_CONFIG_ENTRY_CHANGED` dispatch signal.

Reply:
I think there are a lot of other integration that rely on this pattern.  Since its a race, and may not show up as a test failure, let's add a comment to `async_setup` so we don't break this case unexpectedly.

https://github.com/home-assistant/core/pull/107756